Public/Src/Engine/Dll/EngineSchedule.cs

```diff
@@ -744,6 +744,12 @@ private bool ShouldSerializeOptimizationDataStructurePostExecution()
             IEnumerable<string> nonScrubbablePaths,
             ITempDirectoryCleaner tempCleaner)
         {
+            if (mountPathExpander != null)
+            {
+                // We don't scrub mounts explicitly marked as non-scrubbable
+                nonScrubbablePaths = nonScrubbablePaths.Concat(mountPathExpander.GetNonScrubbableRoots().Select(p => p.ToString(scheduler.Context.PathTable)));

```

Comment:
>nonScrubbablePaths =  [](start = 16, length = 21)

This option is essentially unsafe for shared opaque dirs. A user may accidentally flag a mount as non-scrubbable and any shared opaque contained in that mount won't be cleaned up. This means non-deterministic behavior may occur during the build.
So I'd be ok with this change if we explicitly block shared opaques from being declared under such mounts.
If that's too restrictive, we can add an unsafe flag to disable the block, that users can opt into.

Reply:
The reason I'm interested in this feature is: there is a shared opaque in say path ./A, but the folders ./A/B/*.dir should not be removed (the files within could be removed, but the directory must not be removed).

Thus, what Iman and I came up with is put a tracked r/w non-scrubbable mount in ./A/B, since glob mountpoints are no supported. If I understand correctly, this inherently introduces some non-determinism: initially non-needed files within the mountpoint may or may not be present in the first place.

I'm not sure if
> explicitly block shared opaques from being declared under such mounts.

Also means the intersection should not be allowed.

Reply:
Not scrubbing places where shared opaque outputs can be written to is unsafe. I think I understand the scenario you describe, but allowing that will also open the door to unsafe (unintended) behavior. So if there is a shared opaque A where a pip writes A/B/blah, and a non-scrubbable mount A/B, then blah won't be removed. This is unsafe unless you *really* understand what's going on.

I'd rather have a scrubber-specific configuration (under the schedule configuration) to tell the scrubber to not remove empty directories. That should also be flagged as unsafe though.

---
In reply to: [274675774](https://github.com/Microsoft/BuildXL/pull/130#discussion_r274675774) [](ancestors = 274675774)

Reply:
Changed as reuqested. Please let me know if you have any comments